src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,98 @@
+        {
+            this.DownloadFileShareAsync(downloadParameters.ProjectId, buildArtifacts, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+            return Task.CompletedTask;
+        }
+
+        public Task DownloadFileShareAsync(Guid projectId, IEnumerable<BuildArtifact> buildArtifacts, string targetDirectory, IEnumerable<string> minimatchFilters, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                this.DownloadFileShareAsync(projectId, buildArtifact, dirPath, minimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private Task DownloadFileShareAsync(Guid projectId, BuildArtifact artifact, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken, bool isSingleArtifactDownload = true)
+        {
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            var downloadRootPath = artifact.Resource.Data + Path.DirectorySeparatorChar + artifact.Name;
+            return DirectoryCopyWithMiniMatch(downloadRootPath, destPath, this.context, 1, minimatcherFuncs, downloadRootPath);
+        }
+
+        internal static Task DirectoryCopyWithMiniMatch(string sourcePath, string destPath, AgentTaskPluginExecutionContext context, int parallelCount = 1, IEnumerable<Func<string, bool>> minimatchFuncs = null, string minimatchRoot = null)
+        {
+            // If the source path is a file, the system should copy the file to the dest directory directly. 
+            if(File.Exists(sourcePath)) {
+                context.Output(StringUtil.Loc("CopyFileToDestination", sourcePath, destPath));
+                File.Copy(sourcePath, destPath + Path.DirectorySeparatorChar + Path.GetFileName(sourcePath), true);
+                return Task.FromResult<object>(null);
+            }
+
+            var opts = new ParallelOptions() { MaxDegreeOfParallelism = parallelCount };
+
+            // Get the subdirectories for the specified directory.
+            DirectoryInfo dir = new DirectoryInfo(sourcePath);
+
+            // If the destination directory doesn't exist, create it.
+            if (!Directory.Exists(destPath))
+            {
+                Directory.CreateDirectory(destPath);
+            }
+
+            DirectoryInfo[] dirs = dir.GetDirectories();
+                
+            // Get the files in the directory and copy them to the new location.
+            FileInfo[] files = dir.GetFiles();
+            Parallel.ForEach(files, opts, file => {
```

Comment:
>Parallel.ForEach [](start = 12, length = 16)

You don't want to use Parallel.ForEach because it doesn't separate the work evenly across threads. You want to use ActionBlock because the work will be fed into a Queue which will be evenly distributed. Parallel.ForEach has weird logic where it separates first then distributes buckets of work to each thread.